lib/tree.rb

```diff
@@ -16,49 +16,158 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n) for call stack
+  def add(key, value = nil)
+		new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+		end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+	def add_helper(current, new_node)
+		return new_node if current.nil?
+
+		if new_node.key == current.key
+			current = new_node
+		elsif new_node.key < current.key
+			current.left = add_helper(current.left, new_node)
+		elsif new_node.key > current.key
+			current.right = add_helper(current.right, new_node)
+		end
+
+		return current
+	end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n) for call stack
   def find(key)
```

Comment:
👍 The time complexity is O(log n) if the tree is balanced and O(n) if it isn't.  The space complexity is the same.